### PR TITLE
Plugins: Add error boundary

### LINKED ISSUE #9630: Add error boundaries around plugin components and inside api functions.
**Is your feature request related to a problem? Please describe.**
We would also like to see error boundaries around plugins and maybe also inside of API methods, so if a plugin integration malfunctions, the editor (and other plugins) continues to function normally and an error can be shown which plugin caused the error. This also allows non-developer users to see what plugin is not functioning properly.

**Describe the solution you'd like**
Add a `PluginErrorBoundary` component, that will be rendered as a top-level component around plugin elements in `wp.api.registerPlugin`. This error will make sure an error is shown telling the user what Plugin is faulty.

Add `try/catch` clauses inside of public facing API methods to prevent plugins from causing Gutenberg to crash.

### PR DESCRIPTION
**Pull request by Mamaduka**
## Description
PR adds an error boundary around the plugin renderer component.

The editors can handle displaying error messages using the new `onError` prop of `PluginArea`. It's a callback that will receive the plugin's name and an error.

Resolves #9630.

## Todos
- [x] Add error handlers to other editors.
- [x] Decide if we want to include the "Copy error" action in the notice.
- [ ] Add documentation for `onError` prop.

## Testing Instructions
1. Open a post editor.
2. Paste the snippet in the DevTools console.
3. Check that the new error notice is displayed.

## Screenshots 
CleanShot 2022-02-01 at 12 32 46

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

**Review by youknowriad — COMMENTED** [review]
This is very cool and it looks very simple implementation wise. Can we add an e2e test for it maybe?

Also should we implement in in the site editor too?

**Comment by Mamaduka** [comment]
> Can we add an e2e test for it maybe?

Sounds good to me.

> Also should we implement in the site editor too?

I will add handlers to both Site and Widget editors. I just wanted to share the current implementation for feedback.

Do you think we should add the "Copy error" action to the notice?

I checked other error boundaries, and we have mixed UIs. For example, block boundary doesn't include actions, while general editor one does.

**Comment by youknowriad** [comment]
@Mamaduka I think I'd prefer without it personally as in general the copied error is minified... and doesn't add much value. I didn't hear much feedback about that button in the general editor boundary so would love to be proven wrong :)

**Comment by Mamaduka** [comment]
I agree minified error message and trace isn't practical. At least in my experience when they are included in issues.

We can always include that later if there's a need for it.

Thanks again for the feedback, @youknowriad.

**Comment by Mamaduka** [comment]
I think the code is ready for final review.

After the checks are complete, I will push documentation updates to ensure e2e tests aren't flaky. They work locally, but you never know.

File: packages/plugins/src/components/plugin-area/index.js [hunk 1]
```diff
@@ -114,7 +115,12 @@ class PluginArea extends Component {
 						key={ context.name }
 						value={ context }
 					>
-						<Plugin />
+						<PluginErrorBoundary
```
Comment (ntsekouras) [inline packages/plugins/src/components/plugin-area/index.js:118]: Should we also wrap `PluginContextProvider` with the error boundary?

File: packages/plugins/src/components/plugin-error-boundary/index.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+export class PluginErrorBoundary extends Component {
+	constructor() {
+		super( ...arguments );
```
Comment (ntsekouras) [inline packages/plugins/src/components/plugin-error-boundary/index.js]: Why use the deprecated `arguments` instead of passing props?

**Review by ntsekouras — APPROVED** [review]
Cool work here George!

I've left some minor comments but this looks good! Thanks!

**Comment by talldan** [comment]
>  I think I'd prefer without it personally as in general the copied error is minified... and doesn't add much value. I didn't hear much feedback about that button in the general editor boundary so would love to be proven wrong :)

Some discussion here on the general editor boundary, btw - https://github.com/WordPress/gutenberg/issues/34482

I'm not sure I agree that the error messages aren't useful. They can sometimes provide information on where the error is even when minified. Or at least let the developer know which kind of error is happening.

**Comment by Mamaduka** [comment]
Thanks for mentioning the general issue, @talldan. I just subscribed and am happy to help with development as well.

> I'm not sure I agree that the error messages aren't useful. They can sometimes provide information on where the error is even when minified. Or at least let the developer know which kind of error is happening.

I don't have a strong opinion on this, and we can always add the "Copy" action later.

File: packages/plugins/src/components/plugin-error-boundary/index.js (hunk 2 above)
Comment (Mamaduka) [reply packages/plugins/src/components/plugin-error-boundary/index.js]: Habits are hard to deprecate 😅

File: packages/plugins/src/components/plugin-area/index.js (hunk 1 above)
Comment (Mamaduka) [reply packages/plugins/src/components/plugin-area/index.js:118]: I don't think there's a need for it. The plugin renderer component will be the most likely culprit of error in most cases.

File: packages/plugins/src/components/plugin-error-boundary/index.js (hunk 2 above)
Comment (Mamaduka) [reply packages/plugins/src/components/plugin-error-boundary/index.js]: Resolved in [a579a80](https://github.com/WordPress/gutenberg/pull/38397/commits/a579a8048cab156190e336bd41fa5e7c20400546)

**Comment by Mamaduka** [comment]
Let's merge this. Happy to follow up with the "Copy" action button if/when we see a need for it.

**Comment by Mamaduka** [comment]
@bph, I'm not sure if this needs Dev Note. Yes, it improves developer experience but also requires zero actions from them :)

Anyway, happy to write a few words if we want to mention the improvement.

**Comment by bph** [comment]
@Mamaduka I wasn't quite sure, but from what I learned listening to developer trying to work with blocks sometimes the error messages are not helpful, or don't point to the right places to start debugging. So anytime you improve things, it's a small win to be celebrated. It might not warrant an individual post but we will have a "Miscellaneous' DevNote where we gather a few smaller changes. :-) 
_It's entirely possible, that I don't understand the nature of this PR, so s short write up at least would educate me_ 🤣

**Comment by Mamaduka** [comment]
I think it's worth mentioning in Misc.

P.S. Error boundary won't point folks in the right direction. It just prevents one error from crashing the whole editor.

**Pull request merged**